### DISCUSSION TITLE
Assistance with Automation

### DISCUSSION THREAD

**Discussion by Mike-E-angelo** [datalust/seq-tickets#2067] [Help] [answered]
Hi Team Datalust,

Happy New Year out there.  I am happy to say that I still use Seq after two years of basically learning Azure. 😊  I have been learning Bicep and automation lately and part of that has been wrestling with Seq to make it easier to do automated deployments.

Believe it or not, I have everything working pretty well.  The *only* remaining issue is that I have two different API keys that I use, and I am a bit unclear how to proceed and make this part of the template.

I need to create the API keys and then get them back somehow to the automation process. Is this possible somehow?

FWIW I did see this mention:
https://docs.datalust.co/docs/command-line-client#connecting-without-an-api-key

But it escapes me at the moment how I can utilize this in the Bicep template and/or the PowerShell script that is calling it.  For further context, the goal is to place these keys in an Azure Key Vault as secrets.  Not sure if this is possible from Seq CLI but wanted to mention it in case it's possible.

Thank you for any assistance you can provide,
Michael

**Comment by nblumhardt** [comment]
Hi Michael, Happy New Year to you, too!

I think the first thing to consider would be pre-generating the keys and put them in KeyVault, you could then retrieve them and pass them through when creating the API keys with `seqcli`, e.g.:

```
seqcli apikey create -t Test --token="the secret from KeyVault" ....
```

This at least removes the obstacle of getting keys back into KeyVault later in the process.

If the challenge is to invoke `seqcli` itself, perhaps consider running the `datalust/seqcli` container from the script instead?

The instructions in the "connecting without an API key" docs will only be useful if your deployment has already enabled authentication and has access to the initial username and password - is that the case?

If you can share a bit more background around where you're up to we'll be happy to help you to come up with a scheme :-)

Best regards,
Nick

**Reply by Mike-E-angelo** [reply]
Heya Nick!  Thank you for the kind words and thoughts.  That is an awesome suggestion.  Indeed, I am seeing the `--token` property now, and that makes everything more straightforward for me. I was under the impression that the key had to be new/generated each time, which was throwing me for a loop here.

Knowing this, I am wondering if it's possible to create the keys using a script that gets uploaded to the file share under `/data/Init`.  The catch here is that the keys will still be in plain text after getting renamed to `.done` (if I understand correctly), right?  Is there a way to delete the script instead of renaming it?  If so, that might be a winner.  I am doing uploads already for the certificates so the difficult work there is already done. 😄

**Reply by nblumhardt** [reply]
I think the biggest problem with doing it via an init script would be that, when the init script runs, Seq hasn't yet started up in the container as a server. Since `seqcli` works through the Seq API, this would be a bit of a chicken and egg problem...

Glad the token option is useful, though! 👍

**Reply by Mike-E-angelo** [reply]
OK that takes that route off the board, it sounds like.  After that, am I to understand that I can connect to the server vai the `seqcli` command from the CD/CI machine/environment?  Then it would be a matter of running `dotnet tool install` to get seqcli install, then make calls to container instance with Seq on it, creating the keys.  [Sounds like I would need to create the API access key first](https://github.com/datalust/seqcli?tab=readme-ov-file#connecting-without-an-api-key), however. :)

The consideration here is that I have configured all my Seq instances to authenticate via Active Directory, and now wonder if the above would throw a wrench into this. 🤔

**Reply by Mike-E-angelo** [reply, marked as answer]
Alright @nblumhardt I was able to figure it out.  I generate the administrator password + hash in PowerShell and send that to Bicep.  Bicep returns the server Url, and then I call `seqcli create apikey` from PowerShell using the created password, sending the resulting keys to Azure Key Vault.

The only bummer is that I have to create an administrator user/pass instead of being able to use Entra.  I've got this captured in #2068 as a feature request.

All in all, everything works great!  Thank you for all your tremendous effort out there. 🙏
